src/main/java/org/opensearch/security/user/User.java

```diff
@@ -278,4 +320,78 @@ public boolean isServiceAccount() {
+     */
+    public String toSerializedBase64() {
+        String result = this.serializedBase64;
+
+        if (result == null) {
+            this.serializedBase64 = result = Base64Helper.serializeObject(this);
+        }
+
+        return result;
+    }
+
+    /**
+     * Returns a rough estimated byte size of this object. Used for cache size control.
+     */
+    public int estimatedByteSize() {
+        return this.estimatedByteSize;
+    }
+
+    private int calcEstimatedByteSize() {
+        int size = 32;
+        size += estimateStringSize(this.name);
+        size += estimateStringSize(this.requestedTenant);
+        size += this.roles.stream().mapToInt(User::estimateStringSize).sum() + 32;
+        size += this.securityRoles.stream().mapToInt(User::estimateStringSize).sum() + 32;
+        size += this.attributes.entrySet()
+            .stream()
+            .mapToInt((entry) -> estimateStringSize(entry.getKey()) + estimateStringSize(entry.getValue()))
+            .sum() + 32;
+        return size;
+    }
+
+    private static int estimateStringSize(String s) {
+        if (s != null) {
+            return 40 + s.length() * 2;
+        } else {
+            return 0;
+        }
+    }
+
+    void readObject(ObjectInputStream stream) throws InvalidObjectException {
+        // This object is not supposed to directly read in order to keep compatibility with older OpenSearch versions
+        throw new InvalidObjectException("Use org.opensearch.security.user.serialized.User");
+    }
+
+    /**
+     * Used for creating a backwards compatible object that can be used for serialization.
+     */
+    @Serial
+    private static final ObjectStreamField[] serialPersistentFields = {
+        new ObjectStreamField("name", String.class),
+        new ObjectStreamField("roles", Collections.synchronizedSet(Collections.emptySet()).getClass()),
+        new ObjectStreamField("securityRoles", Collections.synchronizedSet(Collections.emptySet()).getClass()),
+        new ObjectStreamField("requestedTenant", String.class),
+        new ObjectStreamField("attributes", Collections.synchronizedMap(Collections.emptyMap()).getClass()),
+        new ObjectStreamField("isInjected", Boolean.TYPE) };
+
+    /**
+     * Creates a backwards compatible object that can be used for serialization
+     */
+    @Serial
+    private void writeObject(ObjectOutputStream out) throws IOException {
+        ObjectOutputStream.PutField fields = out.putFields();
+        fields.put("name", name);
+        fields.put("roles", Collections.synchronizedSet(new HashSet<>(this.roles)));
+        fields.put("securityRoles", Collections.synchronizedSet(new HashSet<>(this.securityRoles)));
+        fields.put("requestedTenant", requestedTenant);
+        fields.put("attributes", Collections.synchronizedMap(new HashMap<>(this.attributes)));
+        fields.put("isInjected", this.isInjected);
+
+        out.writeFields();
```

Comment:
This is not particularly nice, but it is necessary to keep backwards compatibility with nodes on older OpenSearch versions. To a certain extent, I am inclined again to use a different serialization format on clusters with only new nodes in order to get rid of these synchronizedSet() calls.